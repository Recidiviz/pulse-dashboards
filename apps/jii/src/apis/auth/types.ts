// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { AuthClient } from "~auth";
import { AuthorizedUserProfile } from "~auth0-jii";
import { Hydratable } from "~hydration-utils";

export type AuthorizedUserProperties = {
  userProfile: AuthorizedUserProfile & { name?: string };
};

export interface AuthHandler
  extends Hydratable,
    Partial<AuthorizedUserProperties> {
  authClient?: AuthClient;
  getFirebaseToken: () => Promise<string>;
}

export type UnauthorizedState = { status: "unauthorized" };
export type AuthorizedState = {
  status: "authorized";
} & AuthorizedUserProperties;
export type AuthState = UnauthorizedState | AuthorizedState;

export function isAuthorizedState(state: AuthState): state is AuthorizedState {
  return state.status === "authorized";
}
