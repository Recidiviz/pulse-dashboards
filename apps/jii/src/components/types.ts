// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

/**
 * Simplified version of props for React Router Link
 */
export type SimpleLinkProps = {
  children: string;
  to: string;
};

/**
 * Simplified version of props for React Router NavLink.
 * Send end=true to match entire URL for active state, or false to match descendant links
 */
export type SimpleNavLinkProps = SimpleLinkProps & { end: boolean };
