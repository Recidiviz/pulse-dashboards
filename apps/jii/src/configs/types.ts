// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { z } from "zod";

import { ParsedRecord, ResidentRecord } from "~datatypes";

import { EligibilityReport } from "../models/EligibilityReport/interface";
import { residentOpportunitySchemas } from "./residentsOpportunitySchemas";
import { stateCodes } from "./stateConstants";

export type ProfileField = {
  label: string;
  value: string;
  moreInfo?: string;
};

export type RequirementCopy = { criterion: string; ineligibleReason?: string };

export type OpportunityConfig = {
  urlSection: string;
  firestoreCollection: string;
  htmlTitle: string;
  copy: {
    headline: string;
    subheading: string;
    about: {
      title: string;
      body: string;
      linkText: string;
      fullPage: string;
      fullPageHtmlTitle: string;
    };
    requirements: {
      title: string;
      trackedCriteria: Record<string, RequirementCopy>;
      untrackedCriteria: Array<RequirementCopy>;
      staticRequirementsLabel: string;
      linkText: string;
      fullPage: string;
      fullPageHtmlTitle: string;
    };
    nextSteps: {
      title: string;
      body: string;
      linkText: string;
      fullPage: string;
      fullPageHtmlTitle: string;
    };
    menuLabel: string;
    formPreview: {
      title: string;
    };
  };
};

export type IncarcerationOpportunityId = "usMeSCCP";

export type ResidentsConfig = {
  headerProfileFields: Array<ProfileField>;
  incarcerationOpportunities: Partial<
    Record<IncarcerationOpportunityId, OpportunityConfig>
  >;
};

export type StateCode = z.infer<typeof stateCodes>;

export type ResidentOpportunitySchemaMapping =
  typeof residentOpportunitySchemas;

/**
 * Maps opportunity ID to its successfully parsed record format
 */
export type OpportunityRecord<O extends IncarcerationOpportunityId> =
  ParsedRecord<ResidentOpportunitySchemaMapping[O]>["output"];

/**
 * Mapping of opportunity IDs to their compatible EligibilityReport constructors
 */
export type ResidentEligibilityReportMapping = {
  [Id in IncarcerationOpportunityId]: new (
    ...args: [
      ResidentRecord["output"],
      OpportunityConfig,
      OpportunityRecord<Id> | undefined,
    ]
  ) => EligibilityReport;
};

export type StateConfig<Code extends StateCode = StateCode> = {
  stateCode: Code;
  displayName: string;
  urlSlug: string;
};

export type LandingPageConfig = {
  copy: {
    intro: string;
    selectorLabel: string;
    selectorPlaceholder: string;
  };
};

export type StateLandingPageConfig = {
  copy: {
    intro: string;
    selectorLabel: string;
    selectorPlaceholder: string;
    useCases: {
      intro: string;
      examples: Array<{
        icon: string;
        description: string;
      }>;
    };
  };
  connections: {
    [key: string]: { displayName: string; connectionName: string };
  };
};
