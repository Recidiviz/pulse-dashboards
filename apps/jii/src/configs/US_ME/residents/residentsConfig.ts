// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { ResidentsConfig } from "../../types";
import { config } from "../usMeSCCP/config";
import releaseDateInfo from "./releaseDateInfo.md?raw";

export const usMeResidentsConfig: ResidentsConfig = {
  headerProfileFields: [
    {
      label: "Current release date",
      value:
        "{{#if resident.releaseDate}}{{formatFullDate resident.releaseDate}}{{else}}Not available{{/if}}",
      moreInfo: releaseDateInfo,
    },
  ],
  incarcerationOpportunities: {
    usMeSCCP: config,
  },
  home: {
    progress: {
      title: "Your Progress",
    },
    eligibility: {
      title: "Your Eligibility",
    },
    footer: {
      about: {
        title: "About this app",
        body: `This app is made by Recidiviz, a technology nonprofit.  
        It is free to access. The information updates daily.`,
      },
      contact: {
        title: "Contact us",
        body: `1322 Webster St, Suite 402  
        Oakland, CA 94612`,
      },
    },
  },
};
