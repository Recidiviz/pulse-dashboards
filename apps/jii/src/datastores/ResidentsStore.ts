// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import keyBy from "lodash/keyBy";
import { makeAutoObservable, set } from "mobx";

import { ResidentRecord } from "~datatypes";
import { FilterParams } from "~firestore-api";
import { FlowMethod } from "~hydration-utils";

import { DataAPI } from "../apis/data/interface";
import { residentEligibilityReportConstructors } from "../configs/residentEligibilityReportConstructors";
import {
  IncarcerationOpportunityId,
  OpportunityRecord,
  ResidentsConfig,
} from "../configs/types";
import { EligibilityReport } from "../models/EligibilityReport/interface";
import type { RootStore } from "./RootStore";

// because we use a mapped type to ensure key-value agreement, this can't be an ES6 Map
type OpportunityRecordMapping = {
  [O in IncarcerationOpportunityId]?: OpportunityRecord<O>;
};

export class ResidentsStore {
  /**
   * Holds all resident records that have been fetched
   */
  residentsByExternalId: Map<string, ResidentRecord["output"]> = new Map();

  /**
   * Holds all opportunity eligibility records that have been fetched
   */
  residentOpportunityRecordsByExternalId: Map<
    string,
    OpportunityRecordMapping
  > = new Map();

  residentEligibilityReportsByExternalId: Map<
    string,
    Map<IncarcerationOpportunityId, EligibilityReport>
  > = new Map();

  constructor(
    private readonly rootStore: RootStore,
    public readonly config: ResidentsConfig,
  ) {
    makeAutoObservable(this);
  }

  private get apiClient() {
    return this.rootStore.apiClient;
  }

  get userStore() {
    return this.rootStore.userStore;
  }

  get opportunityIdsByUrlSlug(): Map<string, IncarcerationOpportunityId> {
    return new Map(
      Object.entries(this.config.incarcerationOpportunities).map(
        ([id, config]) => {
          return [
            config.urlSlug,
            id as keyof typeof this.config.incarcerationOpportunities,
          ];
        },
      ),
    );
  }

  /**
   * Get a convenient non-nullable ID value; accepts any string,
   * so you should be confident you have a valid slug or be prepared to handle an error.
   */
  opportunitySlugToIdOrThrow(slug: string) {
    const id = this.opportunityIdsByUrlSlug.get(slug);
    if (!id) {
      throw new Error(`No opportunity ID matches url segment ${slug}`);
    }
    return id;
  }

  areAllResidentsPopulated(): boolean {
    // if we started the session on a single resident's page, we might have had one
    // populated already. Seems a safe assumption that the total will always be > 1
    return this.residentsByExternalId.size > 1;
  }
  /**
   * Populates {@link residentsByExternalId} with the API response for all available residents.
   * Will not refetch if data is already populated, unless `forceRefresh` is true
   */
  *populateResidents(
    filters?: Array<FilterParams>,
    forceRefresh = false,
  ): FlowMethod<DataAPI["residents"], void> {
    if (!forceRefresh && this.areAllResidentsPopulated()) return;

    const residents = yield this.apiClient.residents(filters);
    set(this.residentsByExternalId, keyBy(residents, "personExternalId"));
  }

  isResidentPopulated(residentExternalId: string): boolean {
    return this.residentsByExternalId.has(residentExternalId);
  }

  /**
   * Populates {@link residentsByExternalId} with the API response for the resident
   * with personExternalId matching `residentExternalId`. Throws if the API request fails.
   * Will not refetch if data is already populated.
   */
  *populateResidentById(
    residentExternalId: string,
  ): FlowMethod<DataAPI["residentById"], void> {
    if (this.isResidentPopulated(residentExternalId)) return;

    const resident = yield this.apiClient.residentById(residentExternalId);
    this.residentsByExternalId.set(resident.personExternalId, resident);
  }

  get residentsByPseudoId(): Map<string, ResidentRecord["output"]> {
    return new Map(
      Object.entries(
        keyBy([...this.residentsByExternalId.values()], "pseudonymizedId"),
      ),
    );
  }

  isResidentWithPseudoIdPopulated(pseudoId: string): boolean {
    return this.residentsByPseudoId.has(pseudoId);
  }

  /**
   * Populates {@link residentsByExternalId} with the API response for the resident
   * with pseudonymizedId matching `residentPseudoId`. Throws if the API request fails.
   * Will not refetch if data is already populated.
   */
  *populateResidentByPseudoId(
    residentPseudoId: string,
  ): FlowMethod<DataAPI["residentById"], void> {
    if (this.isResidentWithPseudoIdPopulated(residentPseudoId)) return;

    const resident = yield this.apiClient.residentByPseudoId(residentPseudoId);
    this.residentsByExternalId.set(resident.personExternalId, resident);
  }

  isResidentEligibilityRecordPopulated(
    residentExternalId: string,
    opportunityId: IncarcerationOpportunityId,
  ): boolean {
    return (
      opportunityId in
      (this.residentOpportunityRecordsByExternalId.get(residentExternalId) ??
        {})
    );
  }

  /**
   * Populates {@link residentOpportunityRecordsByExternalId} with the API response for
   * the provided resident and opportunity IDs. Will populate with `undefined` if no record is found
   * (indicating the resident is ineligible). Will not refetch if data is already populated.
   */
  *populateOpportunityRecordByResidentId(
    residentExternalId: string,
    opportunityId: IncarcerationOpportunityId,
  ): FlowMethod<DataAPI["residentEligibility"], void> {
    if (
      this.isResidentEligibilityRecordPopulated(
        residentExternalId,
        opportunityId,
      )
    )
      return;

    const eligibilityRecord = yield this.apiClient.residentEligibility(
      residentExternalId,
      opportunityId,
    );
    const opportunityMap: OpportunityRecordMapping =
      this.residentOpportunityRecordsByExternalId.get(residentExternalId) ?? {};

    opportunityMap[opportunityId] = eligibilityRecord;

    if (!this.residentOpportunityRecordsByExternalId.has(residentExternalId)) {
      this.residentOpportunityRecordsByExternalId.set(
        residentExternalId,
        opportunityMap,
      );
    }
  }

  isResidentEligibilityReportPopulated(
    residentExternalId: string,
    opportunityId: IncarcerationOpportunityId,
  ): boolean {
    return !!this.residentEligibilityReportsByExternalId
      .get(residentExternalId)
      ?.has(opportunityId);
  }

  /**
   * Populates {@link residentEligibilityReportsByExternalId} with an {@link EligibilityReport} object
   * for the given resident and opportunity, if it does not already exist.
   */
  populateEligibilityReportFromData<OppId extends IncarcerationOpportunityId>(
    opportunityId: OppId,
    resident: ResidentRecord["output"],
    opportunity: OpportunityRecord<OppId> | undefined,
  ) {
    const residentExternalId = resident.personExternalId;

    if (
      this.isResidentEligibilityReportPopulated(
        residentExternalId,
        opportunityId,
      )
    ) {
      return;
    }

    const config = this.config.incarcerationOpportunities[opportunityId];

    if (!config) {
      throw new Error(`Opportunity ${opportunityId} is not configured`);
    }

    const ReportConstructor =
      residentEligibilityReportConstructors[opportunityId];

    const mapping =
      this.residentEligibilityReportsByExternalId.get(residentExternalId) ??
      new Map();

    mapping.set(
      opportunityId,
      new ReportConstructor(resident, config, opportunity),
    );
    if (!this.residentEligibilityReportsByExternalId.has(residentExternalId)) {
      this.residentEligibilityReportsByExternalId.set(
        residentExternalId,
        mapping,
      );
    }
  }

  get stateCode() {
    return this.userStore.stateCode;
  }
}
