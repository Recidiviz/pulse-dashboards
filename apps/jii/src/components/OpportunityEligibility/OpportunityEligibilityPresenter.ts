// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { slugify } from "markdown-to-jsx";
import { makeAutoObservable } from "mobx";

import { stateConfigsByStateCode } from "../../configs/stateConstants";
import { OpportunityConfig } from "../../configs/types";
import { ResidentsStore } from "../../datastores/ResidentsStore";
import { EligibilityReport } from "../../models/EligibilityReport/types";
import { State } from "../../routes/routes";
import { LinkProps } from "../ResidentsLayoutRoute/NavigationMenuPresenter";

export class OpportunityEligibilityPresenter {
  constructor(
    private residentsStore: ResidentsStore,
    private config: OpportunityConfig,
    private eligibilityReport: EligibilityReport,
    private residentPseudoId: string,
  ) {
    makeAutoObservable(this, undefined, { autoBind: true });
  }

  get title() {
    return this.eligibilityReport.name;
  }

  get status() {
    return this.eligibilityReport.status;
  }

  private get linkParams() {
    return {
      stateSlug: stateConfigsByStateCode[this.residentsStore.stateCode].urlSlug,
      opportunitySlug: this.config.urlSlug,
      personPseudoId: this.residentPseudoId,
    };
  }

  get requirementsContent() {
    const { status, requirements: sections } = this.eligibilityReport;

    if (status.value === "NA") return;

    const {
      requirements: {
        summary: { heading },
        fullPage,
      },
    } = this.config;

    const { linkText, urlSlug } = fullPage ?? {};
    return {
      id: slugify(heading),
      heading,
      sections,
      linkText,
      linkUrl:
        urlSlug &&
        State.Resident.Eligibility.Opportunity.InfoPage.buildPath({
          ...this.linkParams,
          pageSlug: urlSlug,
        }),
    };
  }

  get additionalSections() {
    return this.eligibilityReport.enabledSections.map((sectionConfig) => {
      const {
        summary: { heading, body },
        fullPage: { urlSlug, linkText },
      } = sectionConfig;
      return {
        id: slugify(heading),
        heading,
        body,
        linkText,
        linkUrl: State.Resident.Eligibility.Opportunity.InfoPage.buildPath({
          ...this.linkParams,
          pageSlug: urlSlug,
        }),
      };
    });
  }

  get tableOfContentsLinks(): Array<LinkProps> {
    const links = [
      ...this.additionalSections.map((section) => ({
        children: section.heading,
        to: `#${section.id}`,
      })),
    ];

    if (this.requirementsContent) {
      links.unshift({
        children: this.requirementsContent.heading,
        to: `#${this.requirementsContent.id}`,
      });
    }

    return links;
  }

  get pageBackgroundStyle() {
    let colors;

    switch (this.status.value) {
      case "ELIGIBLE":
        colors = ["rgba(0, 196, 157, 0.08)", "rgba(255, 255, 255, 0.08)"];
        break;
      case "ALMOST_ELIGIBLE":
        colors = ["rgba(252, 206, 68, 0.2)", "rgba(255, 255, 255, 0.2)"];
        break;
      case "INELIGIBLE":
        colors = ["rgba(0, 77, 72, 0.06)", "rgba(255, 255, 255, 0.06)"];
        break;
    }

    return colors
      ? `linear-gradient(
      180deg,
      ${colors[0]} 0%,
      ${colors[1]} 100%
    )`
      : "none";
  }

  get highlights() {
    if (this.eligibilityReport.highlights.length) {
      return this.eligibilityReport.highlights;
    }
    return undefined;
  }
}
