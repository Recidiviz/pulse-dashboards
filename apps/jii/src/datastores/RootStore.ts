// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { captureException } from "@sentry/react";
import { makeObservable, onReactionError } from "mobx";

import { isOfflineMode, isTestEnv } from "~client-env-utils";
import { FlowMethod } from "~hydration-utils";

import { ApiClient } from "../apis/data/ApiClient";
import { DataAPI } from "../apis/data/interface";
import { OfflineAPIClient } from "../apis/data/OfflineAPIClient";
import { StateCode } from "../configs/types";
import { ResidentsStore } from "./ResidentsStore";
import { UiStore } from "./UiStore";
import { UserStore } from "./UserStore";

// global error handling for Mobx reactions
onReactionError((error) => {
  captureException(error);
});

export class RootStore {
  /**
   * Will be an offline client in offline and test modes,
   * and an external API client otherwise.
   */
  apiClient: DataAPI;

  residentsStore?: ResidentsStore;

  userStore: UserStore;

  uiStore: UiStore;

  // for convenience, this is a constant while we only have one state onboarded
  readonly stateCode: StateCode = "US_ME";

  constructor() {
    makeObservable(this, {
      populateResidentsStore: true,
      residentsStore: true,
    });

    this.userStore = new UserStore(this);

    this.uiStore = new UiStore();

    this.apiClient = this.createApiClient();
  }

  /**
   * Constructs and returns an appropriate API client based on the environment
   * (e.g. one that returns fixture data in offline mode). Not intended to be
   * MobX-reactive, since it creates a brand-new instance and discards any state
   * the previous instance might have held.
   */
  private createApiClient(): DataAPI {
    if (isOfflineMode() || isTestEnv()) {
      return new OfflineAPIClient(this);
    } else {
      // eslint-disable-next-line @typescript-eslint/no-this-alias
      const store = this;
      const externals = {
        // these are getters rather than static properties so that the API client
        // can can observe changes to these values rather than getting a one-time snapshot
        get stateCode() {
          return store.stateCode;
        },
        get authClient() {
          return store.userStore.authClient;
        },
        get config() {
          return store.residentsStore?.config;
        },
      };

      return new ApiClient(externals);
    }
  }

  *populateResidentsStore(): FlowMethod<DataAPI["residentsConfig"], void> {
    const config = yield this.apiClient.residentsConfig();

    this.residentsStore = new ResidentsStore(this, config);
  }
}
