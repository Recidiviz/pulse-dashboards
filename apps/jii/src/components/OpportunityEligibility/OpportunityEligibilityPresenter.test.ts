// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { waitFor } from "@testing-library/react";
import { set } from "mobx";

import { outputFixture, usMeResidents, usMeSccpFixtures } from "~datatypes";

import { residentsConfigByState } from "../../configs/residentsConfig";
import { IncarcerationOpportunityId } from "../../configs/types";
import { ResidentsStore } from "../../datastores/ResidentsStore";
import { RootStore } from "../../datastores/RootStore";
import { UsMeSCCPEligibilityReport } from "../../models/EligibilityReport/UsMe/UsMeSCCPEligibilityReport";
import { OpportunityEligibilityPresenter } from "./OpportunityEligibilityPresenter";

let residentsStore: ResidentsStore;
let presenter: OpportunityEligibilityPresenter;
const opportunityId: IncarcerationOpportunityId = "usMeSCCP";
const resident = outputFixture(usMeResidents[0]);
const stateConfig = residentsConfigByState.US_ME;
const oppConfig = stateConfig.incarcerationOpportunities[opportunityId];

beforeEach(() => {
  residentsStore = new ResidentsStore(new RootStore(), stateConfig);
  presenter = new OpportunityEligibilityPresenter(
    residentsStore,
    resident.personExternalId,
    opportunityId,
    oppConfig,
  );
});

describe("hydration", () => {
  test("needs hydration", () => {
    expect(presenter.hydrationState.status).toBe("needs hydration");
  });

  test("already hydrated", () => {
    set(
      residentsStore.residentsByExternalId,
      resident.personExternalId,
      resident,
    );
    set(
      residentsStore.residentOpportunityRecordsByExternalId,
      resident.personExternalId,
      {
        [opportunityId]: outputFixture(
          usMeSccpFixtures.fullyEligibleHalfPortion,
        ),
      },
    );
    set(
      residentsStore.residentEligibilityReportsByExternalId,
      resident.personExternalId,
      new Map([
        [
          opportunityId,
          new UsMeSCCPEligibilityReport(
            resident,
            oppConfig,
            outputFixture(usMeSccpFixtures.fullyEligibleHalfPortion),
          ),
        ],
      ]),
    );

    expect(presenter.hydrationState.status).toBe("hydrated");
  });

  test("hydrate", async () => {
    expect(presenter.hydrationState.status).toBe("needs hydration");

    presenter.hydrate();

    expect(presenter.hydrationState.status).toBe("loading");

    await waitFor(() =>
      expect(presenter.hydrationState.status).toBe("hydrated"),
    );
  });
});

test("about content", () => {
  expect(presenter.aboutContent).toMatchSnapshot();
});

describe("after hydration", () => {
  beforeEach(async () => {
    await presenter.hydrate();
  });

  test("requirements content", () => {
    expect(presenter.requirementsContent).toMatchInlineSnapshot(`
      {
        "linkText": "Get details about each requirement",
        "linkUrl": "/eligibility/sccp/requirements",
        "sections": [
          {
            "label": "Requirements you've met",
            "requirements": [
              {
                "criterion": "Served 2/3 of your sentence",
              },
              {
                "criterion": "No Class A or B discipline in past 90 days",
              },
              {
                "criterion": "Current custody level is Community",
              },
              {
                "criterion": "No unresolved detainers, warrants or pending charges",
              },
            ],
          },
          {
            "label": "Requirements you haven't met yet",
            "requirements": [
              {
                "criterion": "Fewer than 30 months remaining on your sentence",
                "ineligibleReason": "You'll meet this requirement on May 16, 2022",
              },
            ],
          },
          {
            "label": "Check with your case manager to see if you’ve met these requirements",
            "requirements": [
              {
                "criterion": "Have a safe and healthy place to live for the entire time you are on SCCP",
              },
              {
                "criterion": "Have a plan for supporting yourself – getting a job, going to school, or receiving Social Security or disability benefits",
              },
              {
                "criterion": "Completed required programs, following your case plan, and showing positive change",
              },
            ],
          },
        ],
      }
    `);
  });

  test("next steps content", () => {
    expect(presenter.nextStepsContent).toMatchSnapshot();
  });
});
