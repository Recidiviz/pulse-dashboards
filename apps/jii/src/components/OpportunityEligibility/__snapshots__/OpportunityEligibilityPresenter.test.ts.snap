// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`eligible resident > additional sections content 1`] = `
[
  {
    "body": "While you’re on SCCP, the following rules and conditions apply:

<dl>
<dt>Good Time</dt>
<dd>You still earn good time.</dd>
<dt>Housing</dt>
<dd>You must live in Maine.</dd>
<dt>Employment</dt>
<dd>You must have a job or be looking for one, or be in school. If you qualify for Social Security or disability benefits, you may be able to use those instead.</dd>
<dt>Supervision</dt>
<dd>You have to meet regularly with a probation officer.</dd>
</dl>
",
    "heading": "Quick facts",
    "linkText": "Learn more about how the program works",
    "linkUrl": "/maine/anonres001/eligibility/sccp/about",
  },
  {
    "body": "If you are eligible for SCCP, your case manager will give you an <FormPreview linkText="application packet" /> to fill out, which includes several parts:

<dl>
<dt>Application</dt>
<dd>In the application, you’ll need to outline your release plan, including: your home plan, employment / education plan, treatment (if applicable), and your support system.</dd>
<dt>Essay</dt>
<dd>You’ll have to write an essay describing the path you’re on, the first steps you’ll take when released, and how you plan to do well on SCCP.</dd>
<dt>Paperwork</dt>
<dd>You’ll sign paperwork agreeing to the conditions of the program. You’ll also need to get a warrantless search agreement signed by the person whose place you’ll be living in.</dd>
<dt>Supervision</dt>
<dd>You have to check in regularly with a probation officer while on SCCP.</dd>
</dl>
",
    "heading": "How to apply",
    "linkText": "Learn more about the application process",
    "linkUrl": "/maine/anonres001/eligibility/sccp/application-process",
  },
]
`;

exports[`eligible resident > summary content 1`] = `
{
  "body": "- SCCP is sometimes called “home confinement” or “supervised release”
- You can spend the last part of your sentence under supervision in the community
- SCCP can help you adjust to being back in the community with structured support
",
  "heading": "About the program",
}
`;

exports[`ineligible resident > additional sections content 1`] = `
[
  {
    "body": "While you’re on SCCP, the following rules and conditions apply:

<dl>
<dt>Good Time</dt>
<dd>You still earn good time.</dd>
<dt>Housing</dt>
<dd>You must live in Maine.</dd>
<dt>Employment</dt>
<dd>You must have a job or be looking for one, or be in school. If you qualify for Social Security or disability benefits, you may be able to use those instead.</dd>
<dt>Supervision</dt>
<dd>You have to meet regularly with a probation officer.</dd>
</dl>
",
    "heading": "Quick facts",
    "linkText": "Learn more about how the program works",
    "linkUrl": "/maine/anonres999/eligibility/sccp/about",
  },
]
`;
